cluster/assemblestate/devices.go

```diff
@@ -0,0 +1,256 @@
+package assemblestate
+
+import (
+	"time"
+)
+
+// DeviceQueryTracker manages device identity information and query/response
+// lifecycle for unknown devices in the cluster.
+type DeviceQueryTracker struct {
+	// responses is used to send notifications that indicate that we are able to
+	// respond to a peer's query for device information.
+	responses chan struct{}
+
+	// queries is used to send notifications that indicate that we should query
+	// a peer for device information.
+	queries chan struct{}
+
+	// inflight keeps track of queries for device information that we have
+	// in-flight. We keep track of a mapping of device RDT to time at which the
+	// query was sent.
+	inflight map[DeviceToken]time.Time
+
+	// timeout is the amount of time we will wait before sending a query for
+	// device information again.
+	timeout time.Duration
```

Comment:
nitpick: more of a `queryPeriod` than a timeout since we're not waiting for a query to timeout, we're waiting for some time between queries (but it might be a perfectly normal usage I'm just not familiar with so feel free to ignore)

Reply:
That's fair, I guess I looked at it differently but I see your point :+1: